### PR TITLE
Allowing the first dataset in a list to have more dimensions.

### PR DESCRIPTION
**Pull request by alxmrs**
Allowing users to incorporate multiple datasets that don't have the same number of dimensions, so long as the first dataset is the biggest.

### CODE REVIEW

File: xarray_beam/_src/core.py [hunk 1]
```diff
@@ -430,7 +432,8 @@ def _key_to_chunks(self, key: Key) -> Iterator[Tuple[Key, DatasetOrDatasets]]:
     results = []
     for ds in self._datasets:
       dataset = ds if key.vars is None else ds[list(key.vars)]
```
Comment (alxmrs) [inline xarray_beam/_src/core.py:434]: @shoyer Currently, for the case where the first ds has more dimensions and we split (test case on line 470), this line fails. Specifically, the `key.vars` will index a variable that doesn't exist in the dataset. 

I think I've forgotten the solution we spoke about in VC this morning. Any ideas for how we can address this? Did I understand the new constraint correctly?

File: xarray_beam/_src/core_test.py [hunk 2]
```diff
@@ -456,15 +456,22 @@ def test_validate(self):
...
         ),
     ):
       test_util.EagerPipeline() | xbeam.DatasetToChunks(
-          [dataset, dataset.drop_dims('z')]
+          [dataset.drop_dims('z'), dataset], chunks={'x': 1}
       )
 
+    try:
+      test_util.EagerPipeline() | xbeam.DatasetToChunks(
+        [dataset, dataset.drop_dims('y')], chunks={'x': 1}
```
Comment (shoyer) [inline xarray_beam/_src/core_test.py:470]: `drop_dims` drops all variables that include the dimension, so the first dataset has variables ["foo", "bar", "baz"] and the second only has ["bar", "baz"].

The case we care about supporting is when both datasets have the same data variables, but the some variables on later datasets have fewer dimensions, e.g., `[dataset, dataset.isel(y=0, drop=True)]`

File: xarray_beam/_src/core.py [hunk 3]
```diff
@@ -340,12 +340,14 @@ def _validate(self, dataset, split_vars):
         raise ValueError("dataset list cannot be empty")
     sizes = [ds.sizes for ds in self._datasets]
     if len({tuple(s.items()) for s in sizes}) > 1:
-      raise ValueError(f"inconsistent dataset sizes: {sizes}")
+      if any(len(sizes[0]) < len(s) for s in sizes[1:]):
+        raise ValueError(f"inconsistent dataset sizes, only the first can be greater: {sizes}")
```
Comment (shoyer) [inline xarray_beam/_src/core.py:344]: I'm finding it a little hard to follow these checks written in terms of sizes. I'm also not sure they're right, because you could have a case where the second dataset has fewer dimensions but they are new dimensions.

How about something like this instead?
```python
for ds in self._datasets[1:]:
    for dim, size in ds.sizes.items():
        if dim not in self._first.dims:
            raise ValueError(f"dimension {dim} does not appear on the first dataset")
        if size != self._first.sizes[dim]:
            raise ValueError(f"dimension {dim} has an inconsistent size on different datasets")
```

File: xarray_beam/_src/core.py [hunk 4]
```diff
@@ -340,12 +340,14 @@ def _validate(self, dataset, split_vars):
...
     dv_shapes = [{k: v.shape for k, v in ds.items()} for ds in self._datasets]
     if split_vars and len({tuple(shape) for shape in dv_shapes}) > 1:
-      raise ValueError(
-          f"inconsistent data_var shapes when splitting variables: {dv_shapes}"
-      )
+      if any(len(dv_shapes[0].keys()) < len(dvs.keys()) for dvs in dv_shapes[1:]):
```
Comment (shoyer) [inline xarray_beam/_src/core.py:347]: I actually think we only need to check to make sure variable names are the same on each dataset. It's OK if they have different shapes (that's the entire idea of allowing the first dataset to have more dimensions).

So someting like the following should work:
```
if split_vars:
    for ds in self._datasets:
        if not ds.keys() <= self._first.keys():
            raise ValueError
```

File: xarray_beam/_src/core.py [hunk 5]
```diff
@@ -340,12 +340,14 @@ def _validate(self, dataset, split_vars):
...
+        raise ValueError(
+            f"inconsistent data_var shapes when splitting variables: {dv_shapes}"
+        )
     chunks = [ds.chunks for ds in self._datasets]
     if len({tuple(c.items()) for c in chunks}) > 1:
       raise ValueError(f"inconsistent chunks: {chunks}")
```
Comment (shoyer) [inline xarray_beam/_src/core.py:353]: I think this check could probably be removed. Chunking with Dask is not directly tied to chunking when reading data in Xarray-Beam, and it could be valid to have different Dask chunk sizes on different datasets, especially if the Dask chunks always fit into Beam chunks.

**Review by shoyer — APPROVED** [review]
